ts/components/conversation/composition/CompositionTextArea.tsx

```diff
@@ -91,77 +74,266 @@ function useMembersInThisChat(): Array<SessionSuggestionDataItem> {
+    }
+
+    const username = model.getRealSessionUsername();
+    if (username && username !== nickname) {
+      searchable.push(username.toLowerCase());
+    }
+
+    const isYou = UserUtils.isUsFromCache(id);
+    const display = isYou ? localize('you').toString() : nickname || PubKey.shorten(id);
+
+    if (display && display !== nickname) {
+      searchable.push(display.toLowerCase());
+      if (isYou) {
+        const enYou = localize('you').forceEnglish().toString();
+        if (enYou !== display) {
+          searchable.push(localize('you').forceEnglish().toString().toLowerCase());
+        }
+      }
+    }
+
     return {
-      id: m,
-      display: UserUtils.isUsFromCache(m)
-        ? localize('you').toString()
-        : ConvoHub.use().get(m)?.getNicknameOrRealUsernameOrPlaceholder() || PubKey.shorten(m),
+      id,
+      display,
+      searchable,
     };
   });
 }
 
-function fetchMentionData(
-  query: string,
-  fetchedMembersInThisChat: Array<SessionSuggestionDataItem>
-): Array<SessionSuggestionDataItem> {
-  let overriddenQuery = query;
-  if (!query) {
-    overriddenQuery = '';
+enum PREFIX {
+  USER = '@',
+  EMOJI = ':',
+}
+
+type MentionDetails = {
+  prefix: PREFIX;
+  content: string;
+};
+
+const MENTION_ESCAPE_STRINGS = [' ', '\n'];
+
+function findValidMention(val: string, cursorPosition: number, prefix: PREFIX) {
+  const pos = val.lastIndexOf(prefix);
+  if (pos === -1) {
+    return null;
   }
 
-  return filterMentionDataByQuery(overriddenQuery, fetchedMembersInThisChat);
+  const content = val.substring(pos + 1, cursorPosition);
+
+  if (MENTION_ESCAPE_STRINGS.some(char => content.lastIndexOf(char) !== -1)) {
+    return null;
+  }
+
+  return {
+    content,
+    prefix,
+  };
+}
+
+function getMentionDetails(
+  val: string,
+  cursorPosition: number,
+  maxLength: number
+): MentionDetails | null {
+  if (!val || cursorPosition > val.length || maxLength <= 0) {
+    return null;
+  }
+
+  const startIdx = Math.max(0, cursorPosition - maxLength);
+  const searchableVal = val.slice(startIdx, cursorPosition);
+
+  const userMention = findValidMention(searchableVal, cursorPosition, PREFIX.USER);
+  if (userMention) {
+    return userMention;
+  }
+
+  const emojiMention = findValidMention(searchableVal, cursorPosition, PREFIX.EMOJI);
+  if (emojiMention) {
+    return emojiMention;
+  }
+
+  return null;
+}
+
+function createUserMentionHtml({ id, display }: SessionSuggestionDataItem) {
+  return `${renderToString(
+    <Mention dataUserId={id} key={id} text={display} inComposableElement={true}>
+      <ZeroWidthNode />
+    </Mention>
+  )} `;
 }
 
 export const CompositionTextArea = (props: Props) => {
-  const { draft, setDraft, container, textAreaRef, typingEnabled, onKeyDown } = props;
+  const { draft, initialDraft, setDraft, inputRef, typingEnabled, onKeyDown } = props;
 
   const [lastBumpTypingMessageLength, setLastBumpTypingMessageLength] = useState(0);
```

Comment:
This component is way too big. 
Can you at least take the hooks out (but keep them in the same file):
`const handleSelect = useCallback`
`const popoverContent = useMemo(() =>`
`handleMentionCheck`
`const handleKeyDown = useCallback`

It will make it a bit  more digest